lib/vnet/dns/dns.go

```diff
@@ -0,0 +1,402 @@
+	requestHeader, err := parser.Start(buf)
+	if err != nil {
+		return trace.Wrap(err, "parsing DNS message")
+	}
+	if requestHeader.OpCode != 0 {
+		slog.DebugContext(ctx, "OpCode is not QUERY (0), forwarding.", "opcode", requestHeader.OpCode)
+		return trace.Wrap(s.forward(ctx, slog, buf, responseWriter), "forwarding non-Query DNS message")
+	}
+	question, err := parser.Question()
+	if err != nil {
+		return trace.Wrap(err, "parsing DNS question")
+	}
+	fqdn := question.Name.String()
+	slog = slog.With("fqdn", fqdn, "type", question.Type.String())
+	slog.DebugContext(ctx, "Received DNS question.", "question", question)
+	if question.Class != dnsmessage.ClassINET {
+		slog.DebugContext(ctx, "Query class is not INET, forwarding.", "class", question.Class)
+		return trace.Wrap(s.forward(ctx, slog, buf, responseWriter), "forwarding non-INET DNS query")
+	}
+
+	var result Result
+	switch question.Type {
+	case dnsmessage.TypeA:
+		result, err = s.resolver.ResolveA(ctx, fqdn)
+		if err != nil {
+			return trace.Wrap(err, "resolving A request for %q", fqdn)
+		}
+	case dnsmessage.TypeAAAA:
+		result, err = s.resolver.ResolveAAAA(ctx, fqdn)
+		if err != nil {
+			return trace.Wrap(err, "resolving AAAA request for %q", fqdn)
+		}
+	default:
+		slog.DebugContext(ctx, "Question type is not A or AAAA, forwarding.", "type", question.Type)
+		return trace.Wrap(s.forward(ctx, slog, buf, responseWriter), "forwarding %s DNS query", question.Type)
+	}
+
+	var response []byte
+	switch {
+	case result.NXDomain:
+		slog.DebugContext(ctx, "No match for name, responding with authoritative name error.")
+		response, err = buildNXDomainResponse(buf, &requestHeader, &question)
+	case result.NoRecord:
+		slog.DebugContext(ctx, "Name matched but no record, responding with authoritative non-answer.")
+		response, err = buildEmptyResponse(buf, &requestHeader, &question)
+	case question.Type == dnsmessage.TypeA && result.A != ([4]byte{}):
+		slog.DebugContext(ctx, "Matched DNS A.", "a", tcpip.AddrFrom4(result.A))
+		response, err = buildAResponse(buf, &requestHeader, &question, result.A)
+	case question.Type == dnsmessage.TypeAAAA && result.AAAA != ([16]byte{}):
+		slog.DebugContext(ctx, "Matched DNS AAAA.", "aaaa", tcpip.AddrFrom16(result.AAAA))
+		response, err = buildAAAAResponse(buf, &requestHeader, &question, result.AAAA)
+	default:
+		slog.DebugContext(ctx, "Forwarding unmatched query.")
+		return trace.Wrap(s.forward(ctx, slog, buf, responseWriter), "forwarding unmatched DNS query")
+	}
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	_, err = responseWriter.Write(response)
+	return trace.Wrap(err, "writing DNS response")
+}
+
+// forward forwards a raw DNS message to all upstream nameservers and writes the first response to
+// [responseWriter]. If there are no upstream nameservers, or none of them responds within the timeout, an
+// error is returned. This doesn't do any retries because the downstream resolver is likely to do its own
+// retries.
```

Comment:
I'm a little confused by the upstream/downstream terminology being used here. Aren't we the first nameserver that an outgoing DNS message is processed by, and all other nameservers are "downstream" from us?

Reply:
it basically goes `(client binary) -> (macOS resolver) -> (vnet nameserver) -> (upstream nameserver) -> (who knows)`. I'm saying to the left and towards to user is downstream, to the right and towards "the internet" is upstream, I think that's pretty standard

Reply:
Thanks, I wasn't aware there was a convention in terms of the client vs "internet". 

(Still, I can't shake the feeling that they've chosen the wrong metaphor. By that metaphor, DNS resolution queries travel from downstream to upstream. I guess they have little motorized boats or something).